packages/next/src/server/pipe-readable.ts

```diff
@@ -1,107 +1,115 @@
+    aborted = true
+    finished = true
+    controller.abort()
   }
-  writable.on('close', onClose)
 
-  try {
-    while (true) {
-      const { done, value } = await reader.read()
-      readerDone = done
+  // When the response is closed, we should abort the stream if it hasn't
+  // finished yet.
+  res.once('close', onClose)
 
-      if (done || writableClosed) {
-        break
+  // Create a writable stream that will write to the response.
+  return new WritableStream<Uint8Array>({
+    write: async (chunk) => {
+      // You'd think we'd want to use `start` instead of placing this in `write`
+      // but this ensures that we don't actually flush the headers until we've
+      // started writing chunks.
+      if (!started) {
+        started = true
+        res.flushHeaders()
       }
 
-      if (value) {
-        writable.write(value)
-        writable.flush?.()
+      try {
+        const ok = res.write(chunk)
+
+        // Added by the `compression` middleware, this is a function that will
+        // flush the partially-compressed response to the client.
+        if ('flush' in res && typeof res.flush === 'function') {
+          res.flush()
+        }
+
+        // If the write returns false, it means there's some backpressure, so
+        // wait until it's streamed before continuing.
+        if (!ok) {
+          await new Promise((resolve) => {
+            res.once('drain', resolve)
+          })
+        }
+      } catch (err: any) {
+        res.end()
+
+        throw err
       }
-    }
-  } catch (e) {
-    // If the client disconnects, we don't want to emit an unhandled error.
-    if (!isAbortError(e)) {
-      throw e
-    }
-  } finally {
-    writable.off('close', onClose)
+    },
+    abort: () => {
+      finished = true
+      res.end()
+    },
+    close: () => {
+      finished = true
 
-    // If we broke out of the loop because of a client disconnect, and the
-    // close event hasn't yet fired, we can early cancel.
-    if (!readerDone) {
-      reader.cancel().catch(() => {})
-    }
+      if (res.writableFinished) return
 
-    // If the client hasn't disconnected yet, end the writable so that the
-    // response sends the final bytes.
-    if (waitUntilForEnd) {
-      await waitUntilForEnd
-    }
+      // Create a promise that will resolve once the response has finished.
+      //
+      // See: https://nodejs.org/api/http.html#event-finish_1
+      //
+      const promise = new Promise<void>((resolve) =>
+        res.once('finish', () => resolve())
+      )
+
+      res.end()
+      return promise
+    },
+  })
+}
+
+export async function pipeToNodeResponse(
+  readable: ReadableStream<Uint8Array>,
+  res: ServerResponse
+) {
+  try {
+    // Create a new AbortController so that we can abort the readable if the
+    // client disconnects.
+    const controller = new AbortController()
+
+    const writer = createWriterFromResponse(res, controller)
```

Comment:
There's `signalFromNodeResponse` in `packages/next/src/server/web/spec-extension/adapters/next-request.ts`, which we're having to reimplement here.